content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
+        if "args" in instr:
+            instr["args"] = [ultimate_src(a) for a in instr["args"]]
+```
+
+#### 2.3.2 Key Insight: Phi Argument Propagation
+
+A subtle but important point: we **can** propagate copies into phi arguments in SSA form. This is safe because in SSA, every definition dominates all its uses. If `x.1 = id y.1`, then wherever `x.1` appears (including in phi nodes in successor blocks), we can replace it with `y.1`.
+
+This is a key advantage of performing copy propagation **before** SSA lowering.
+
+### 2.4 Dominator-Tree Copy Elimination {#24-dominator-tree-copy-elimination}
+
+#### 2.4.1 Design
+
+After SSA copy propagation, there may still be `id` instructions that weren't eliminated because they define phi destinations or have other uses. The dominator-tree copy elimination pass walks the dominator tree with a **rename environment**:
+
+```python
+def dom_tree_copy_elimination(func: dict):
+    """Eliminate copies using dominator-tree-based propagation.
+
+    Walks the dominator tree carrying a rename environment. For each `id`
+    instruction, we establish a rename from dest to source and drop the
+    instruction.
+    """
+    def walk(block_name: str, rename: Dict[str, str]):
+        block = bmap[block_name]
+        new_instrs = []
+
+        for instr in block.get("instrs", []):
+            op = instr.get("op")
+
+            # Rewrite arguments using rename map (but NOT phi args)
+            if "args" in instr and op != "phi":
+                instr["args"] = [canon(a, rename) for a in instr["args"]]
+
+            # Handle id instructions
+            if op == "id":
+                dest = instr.get("dest")
+                if dest not in phi_dests:
+                    src = canon(args[0], rename)
+                    rename[dest] = src
+                    continue  # Drop the id instruction
+
+            new_instrs.append(instr)
+
+        block["instrs"] = new_instrs
+
+        # Recurse with copied environment
+        for child in dom_tree.get(block_name, []):
+            walk(child, rename.copy())
+```
+
+#### 2.4.2 Why This is Safe
+
+The key insight is that renames established in a dominator are valid in all dominated blocks. By copying the environment before recursing into children, we ensure that renames from one branch don't leak into sibling branches.
+
+**Example:**
+```
+     B0: x = 5
+    /  \
+   B1   B2
+   y=x  z=x
+```
+
+If we're at B0 and establish `x → 5`, both B1 and B2 see this rename. But renames in B1 don't affect B2 because they receive separate copies of the environment.
+
+#### 2.4.3 Preserving Phi Destinations
+
+We explicitly preserve phi destinations because they are **merge points**. Even if a phi produces `x = phi(y, z)` where `y` and `z` have the same value, we can't eliminate the phi — it defines a new SSA name that may be used later.
+
+### 2.5 Fixed Parallel Copy Sequentialization {#25-fixed-parallel-copy-sequentialization}
+
+#### 2.5.1 The Problem
```

Comment:
This is called the "swap problem," and it is discussed in the 6120 lecture on SSA.